### ISSUE TITLE
play_tone() not working

### ISSUE THREAD

**Issue by caternuson** [adafruit/Adafruit_CircuitPython_CircuitPlayground#40]
Confirmed. See here for more info and sample code:
https://forums.adafruit.com/viewtopic.php?f=58&t=140985

**Comment by sommersoft** [comment]
I can _third_ this with 3.0.2. Used...
```
>>> for i in range(50):
...     tone = 300 + (i * 10)
...     print("Tone:", tone)
...     cpx.play_tone(tone, 0.5)
...
```
… which has a noticeable droop in the ~540Hz to ~690Hz range. 

Additionally, I ran this with a factory fresh CPX with 2.2.0. The tones increase accordingly with above code. May need to debug `audioio.AudioOut` and/or `audioio.RawSample`.

**Comment by caternuson** [comment]
Agree, I think it's going to be something lower level than CPX. I can recreate with this:
```python
import array
import math
import time

import board
import digitalio
import audioio

speaker_enable = digitalio.DigitalInOut(board.SPEAKER_ENABLE)

# Generator thing
def sine_sample(length):
    tone_volume = (2 ** 15) - 1
    shift = 2 ** 15
    for i in range(length):
        yield int(tone_volume * math.sin(2*math.pi*(i / length)) + shift)

# Generate the sample
length = 100
sine_wave = array.array("H", sine_sample(length))
sample = audioio.AudioOut(board.SPEAKER)
sine_wave_sample = audioio.RawSample(sine_wave)

# Play tone
frequency = 330
sample_rate = int(len(sine_wave) * frequency)
speaker_enable = True
sine_wave_sample.sample_rate = sample_rate
sample.play(sine_wave_sample, loop=True)
time.sleep(1)
sample.stop()
sample.deinit()
speaker_enable = False
```

**Comment by ladyada** [comment]
ooooh ok maybe this is a @tannewt @dhalbert level issue. will re-assign!

**Comment by tannewt** [comment]
I've opened https://github.com/adafruit/circuitpython/issues/1196 to continue this investigation deeper. I'll also take a quick look now. Thanks!

**Comment by tannewt** [comment]
This is actually an issue with both.

The limit of the DAC is 350ksps which we weren't doing because the input clock was too slow. With the fix for https://github.com/adafruit/circuitpython/issues/1196 660hz will work.

However, this library will start to crash when play_tones frequency is over 3500 ( * 100 is 350ksps) because it is now checked correctly.

So, I'm reopening this as well. The sample length should be reduced when the frequency is more than 3500 in order to keep the sample rate below the limit.

**Comment by caternuson** [comment]
> The sample length should be reduced when the frequency is more than 3500 in order to keep the sample rate below the limit.
Take it you mean this needs to be done here in this lib? Instead of hardwired length of 100?

@ladyada should this be re-re-assigned?

**Comment by ladyada** [comment]
i dont know 😂
scott, do ya want carter to limit this here?

**Comment by tannewt** [comment]
Yes please @caternuson ! Thank you!

**Comment by caternuson** [comment]
Closing. Fixed with #41.
